### PR TITLE
Feat/export base widget

### LINKED ISSUE #58: Facets in a select or way to ng-template ng-ais-refinement-list
Hello,

I want to be able to put my facets in a select multiple, or at least in a ng-template like the results, is there already a way to do this ?

Regards,

### PR DESCRIPTION
**Pull request by iam4x**
Fixes #58 #40 

With this PR, we are now exporting the usage of the intern `BaseWidget` class. This allow to create custom widgets on top of InstantSearch.js connectors in a easy way for Angular.

### CODE REVIEW

**Comment by iam4x** [comment]
Deploy preview for *angular-instantsearch* ready!

Built with commit 3c1bfab9b4484e6db92dcf8c4d42faa9cc590899

https://deploy-preview-59--angular-instantsearch.netlify.com

**Comment by vvo** [comment]
deploy is not live weirdly

**Comment by iam4x** [comment]
I do not have a `index.html` for netlify, you can see the custom widget here: https://deploy-preview-59--angular-instantsearch.netlify.com/dev-novel/?selectedStory=CustomWidgets.MenuSelect

**Comment by iam4x** [comment]
Merging now, will iterate on the documentation with the feedback.

File: doc/custom-widgets.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
+## Customize widgets
+
+Angular InstantSearch comes with a set of UI widgets. Some of them have options to modify the HTML output (`<ng-template>`).
```
Comment (vvo) [inline doc/custom-widgets.md]: From reading the introduction I do not understand what is the purpose of the guide. Maybe turn it this way:

```txt
Angular InstantSearch comes with a set of UI widgets. Some of them have options to modify the HTML output (`<ng-template>`). But sometime that is not sufficient and you want to completely customise the rendering of a widget.

For this you can use our BaseWidget class.
```

File: doc/custom-widgets.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+Every of these widgets are built using the [InstantSearch.js connectors API](https://community.algolia.com/instantsearch.js/v2/connectors.html) which contains the logic of the widgets without the rendering.
+
+### Introduce BaseWidget class
```
Comment (vvo) [inline doc/custom-widgets.md]: Remove Introduce

File: doc/custom-widgets.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+The `BaseWidget` class helps you to create new widgets using the InstantSearch.js connectors with Angular.
+
+It encapsulate the logic to maintain the state of the widget in sync with the search and handles for you the initialisation and the dispose of the widget on Angular life cycle hooks.
+
+Let's go step by step on how to write a custom widget using the `BaseWidget` class.
+
+## Custom Menu Select widget
+
+In this example we will create a new custom widget using the `BaseWidget` class and the [connectMenu](https://community.algolia.com/instantsearch.js/v2/connectors/connectMenu.html) InstantSearch.js connector.
```
Comment (vvo) [inline doc/custom-widgets.md]: Add

```txt
The default menu widget renders a list of links, but we would like to render it as a select element instead.
```

File: doc/custom-widgets.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+To know the options used by the connector we can refer to the [InstantSearch.js connectors API documentation](https://community.algolia.com/instantsearch.js/v2/connectors.html).
+
+In our case, we will use the `connectMenu` which accepts multiple options but for simplicity we will only use the `attributeName` option which is the only one mandatory.
+
+(The `attributeName` is the name of the attribute for faceting)
+
+```ts
+import { Component, Inject, PLATFORM_ID } from '@angular/core';
+import { BaseWidget, NgAisInstance } from 'angular-instantsearch';
+import { connectMenu } from 'instantsearch.js/es/connectors';
```
Comment (vvo) [inline doc/custom-widgets.md]: As soon as we do this, it means we are introducing a new public API (connectors). I am fine with it but now we will have to be extra careful with it (and also at some point be able to remove that /es/), @samouss I remember a previous discussion on how to remove this /es/, am I right? Any pointers?

File: doc/custom-widgets.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  public ngOnInit() {
+    this.createWidget(connectMenu, { attributeName: 'categories' });
+    super.ngOnInit();
+  }
+}
+```
+
+And voilà, your widget is connected to InstantSearch.js and the state of the search it self!
```
Comment (vvo) [inline doc/custom-widgets.md]: And voilà => That's it!

Add

```txt
Now let's update the rendering of this widget
```

File: doc/custom-widgets.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  public ngOnInit() {
+    this.createWidget(connectMenu, { attributeName: 'categories' });
+    super.ngOnInit();
+  }
+}
+```
+
+You can now use your directive `<ng-ais-menu-select></ng-ais-menu-select>` to display your custom MenuSelect widget!
+
+See your freshly crafted MenuSelect in action [here](http://angular-instantsearch.netlify.com/dev-novel/?selectedStory=CustomWidgets.MenuSelect).
```
Comment (vvo) [inline doc/custom-widgets.md]: We have a live example of this newly created MenuSelect in our [widget showcase].

**Comment by vvo** [comment]
Sorry had some comment but was slow to add them, now done

File: doc/custom-widgets.md (hunk 4 above)
Comment (iam4x) [reply doc/custom-widgets.md]: The `/es/` is needed for having a complete tree-shaking and the possibility to target the exact files.

We document the `/es/` prefix on IS.js here: https://community.algolia.com/instantsearch.js/v2/guides/usage.html#optimize-your-build-with-tree-shaking

**Comment by iam4x** [comment]
Thanks for the comments @vvo 👍

**Pull request merged**